### PR TITLE
Add support for version 18 of React

### LINKED ISSUE #7142: Core - Add support for React 18
React 18 has been [released](https://reactjs.org/blog/2022/03/29/react-v18.html), code will have to be adjusted to that it is officially supported.

### Checklist
- [x] https://github.com/patternfly/patternfly-react/issues/6919
- [x] Update `@types/react` and `@types/react-dom` (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56210).
- [x] #7195
- [x] #7216
- [x] #7217
- [x] #7218
- [x] #7282
- [x] #7363
- [x] #7402
- [x] Add `PropsWithChildren` types to components with children
- [ ] Apply migrations from [upgrade guide](https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html).
  - [x] #8027
  - [ ] https://github.com/patternfly/patternfly-org/pull/3203
- [ ]  Check for incompatible dependencies and upgrade if necessary.
  - [x] #7706
  - [x] #7826
  - [x] #8577
- [x] #7303
- [x] #7821
- [x] #7827

### PR DESCRIPTION
**Pull request by jonkoops**
Works towards closing #7142

### CODE REVIEW

File: packages/react-docs/package.json
Comment (jonkoops) [inline packages/react-docs/package.json:44]: I removed these dependencies as they are already defined in the root package.

File: .eslintrc.json
Comment (jonkoops) [inline .eslintrc.json:27]: This allows the React ESLint plugin to [detect](https://github.com/yannickcr/eslint-plugin-react#configuration) the appropriate React version automatically.

**Review by wise-king-sullyman — COMMENTED** [review]
Everything here looks good to me, I'm ready to approve once you resolve the merge conflicts!

**Comment by jonkoops** [comment]
Sorry for the delay, I was on vacation so I had no time to look at this. @wise-king-sullyman I've rebased the PR.

**Comment by wise-king-sullyman** [comment]
@jonkoops it looks like there are unit tests failing now, based on some quick research though I think those failures should be resolved if you bump our React Testing Library version.

**Comment by jonkoops** [comment]
@wise-king-sullyman did another rebase and updated the version of React Testing Library. All is looking green now :heavy_check_mark:

**Review by wise-king-sullyman — APPROVED** [review]

**Review by tlabaj — APPROVED** [review]
LGTM!  Thank you!

**Pull request merged**
